### PR TITLE
Gateway 3249

### PR DESCRIPTION
**Pull request by bhumphrey**
- Created new module for DocQueryErrorsCodes
- Refactored Fanout to move the condition when there are not matching patients to closer to generation of the error

### CODE REVIEW

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/AggregationService.java [hunk 1]
```diff
@@ -75,6 +78,10 @@
     private PixRetrieveBuilder pixRetrieveBuilder;
     private StandardOutboundDocQueryHelper standardOutboundDocQueryHelper;
 
+    private static final OutboundResponseProcessor NULL_OutboundResponseProcessor = null;
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/AggregationService.java]: The code smell is that the method accepts a bunch of params that could be null.  This is a half-measure.  Can you create another method in the target class that doesn't take the params, and delegates with nulls internally?

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java [hunk 2]
```diff
@@ -50,27 +53,29 @@
...
     private AggregationStrategy strategy;
-   
+    private AggregationService fanoutService;
+    
+
     /**
      * Add default constructor that is used by test cases Note that implementations should always use constructor that
      * takes the executor services as input.
      */
     public StandardOutboundDocQuery() {
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java:65]: Recommend you do this as

``` java
this(new AggregationStrategy(), new AggregationService());
```

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java [hunk 3]
```diff
@@ -96,23 +99,30 @@ public StandardOutboundDocQuery(ExecutorService e, ExecutorService le) {
...
-        aggregate.setAdhocQueryRequest(adhocQueryRequest);
-        aggregate.setAssertion(assertion);
-        aggregate.setTargets(targets);
-        OutboundDocQueryOrchestratable request= new OutboundDocQueryOrchestratable(new OutboundDocQueryAggregator(), assertion, NhincConstants.DOC_QUERY_SERVICE_NAME, adhocQueryRequest);
-  
+
+        List<OutboundOrchestratable> aggregateRequests = fanoutService
+                .createChildRequests(adhocQueryRequest, assertion, targets);
+
+        if (aggregateRequests.size() == 0) {
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java]: or aggregateRequests.isEmpty()

File: Product/Production/Services/DocumentQueryCore/src/test/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQueryTest.java [hunk 4]
```diff
@@ -45,28 +47,29 @@ public void testrespondingGatewayCrossGatewayQueryforNullEndPoint() {
...
 
         NhinTargetCommunitiesType targets = createNhinTargetCommunites();
         AdhocQueryResponse response = entitydocqueryimpl.respondingGatewayCrossGatewayQuery(adhocQueryRequest,
                 assertion, targets);
 
-        ArgumentCaptor<OutboundDocQueryAggregate> aggregate = ArgumentCaptor.forClass(OutboundDocQueryAggregate.class);
 
-        verify(strategy).execute(aggregate.capture());
+        verify(service).createChildRequests(eq(adhocQueryRequest), eq(assertion), eq(targets));
+    //    verify(strategy).execute(aggregate);
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/test/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQueryTest.java]: Commented out code??

File: Product/SoapUI_Test/RegressionSuite/DocQueryErrorCodes/pom.xml [hunk 5]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.connectopensource</groupId>
+        <artifactId>SoapUI_Test</artifactId>
+        <version>4.0.0-SNAPSHOT</version>
+    </parent>
+    <artifactId>DocQueryErrorCodes</artifactId>
+    <packaging>jar</packaging>
```
Comment (tparnell) [inline Product/SoapUI_Test/RegressionSuite/DocQueryErrorCodes/pom.xml:10]: Why jar packaging?  Isn't pom enough?

Comment (bhumphrey) [inline Product/SoapUI_Test/RegressionSuite/DocQueryErrorCodes/pom.xml:10]: exact reason, not sure right now. But I believe it has to do with test resource processing.

**Pull request merged**